PyRoute/DeltaReduce.py

```diff
@@ -0,0 +1,325 @@
+"""
+Created on May 23, 2023
+
+The core reduction engine used by delta debugging.
+Modify this class to add different reduction passes.
+
+@author: CyberiaResurrection
+"""
+import functools
+import logging
+import math
+
+from PyRoute.DeltaDictionary import DeltaDictionary
+from PyRoute.DeltaGalaxy import DeltaGalaxy
+from PyRoute.HexMap import HexMap
+from PyRoute.SpeculativeTrade import SpeculativeTrade
+from PyRoute.StatCalculation import StatCalculation
+from PyRoute.SubsectorMap2 import GraphicSubsectorMap
+
+
+class DeltaReduce:
+
+    def __init__(self, sectors, args, interesting_line=None, interesting_type=None):
+        assert isinstance(sectors, DeltaDictionary), "Sectors object must be an instance of DeltaDictionary"
+        self.sectors = sectors
+        self.args = args
+        # Interesting_line allows the caller to tighten the definition of interesting by requiring a specific
+        # string to appear in the exception message
+        # Interesting_type requires an exception type to contain a specific string
+        # If both are defined, they both have to match for the result to be interesting
+        self.interesting_line = interesting_line
+        self.interesting_type = interesting_type
+        self.logger = logging.getLogger('PyRoute.Star')
+        logging.disable(logging.WARNING)
+
+    def is_initial_state_interesting(self):
+        sectors = self.sectors
+        args = self.args
+
+        interesting, _ = self._check_interesting(args, sectors)
+
+        if not interesting:
+            raise AssertionError("Original input not interesting - aborting")
+
+    def is_initial_state_uninteresting(self):
+        sectors = self.sectors
+        args = self.args
+
+        interesting, msg = self._check_interesting(args, sectors)
+
+        if interesting:
+            raise AssertionError(msg)
+
+    def reduce_sector_pass(self, singleton_only=False):
+        segment = self.sectors.sector_list()
+
+        # An interesting single-element list is 1-minimal by definition
+        if 2 > len(segment):
+            return
+
+        num_chunks = len(segment) if singleton_only else 2
+        short_msg = None
+        best_sectors = self.sectors
+        singleton_run = singleton_only
+
+        while num_chunks <= len(segment):
+            chunks = self.chunk_lines(segment, num_chunks)
+            remove = []
+            msg = "# of lines: " + str(len(best_sectors.lines)) + ", # of chunks: " + str(num_chunks) + ", # of sectors: " + str(len(segment))
+            self.logger.error(msg)
+
+            for i in range(0, num_chunks):
+                threshold = i + (len(remove) if 2 == num_chunks else 0)
```

Comment:
If the first of two chunks has been dropped due to uninterestingness, we have already checked the second chunk and don't need to do so again.